---
title: "Pathway enrichment analysis"
author: "Pascal Lafrenz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Pathway enrichment analysis

A possible downstream analysis in R is an over-representation analysis (ORA) to determine whether genes from specific pathways (e.g. KEGG pathways) are present more than expected in our network.

For that we need the following inputs: a matrix to evaluate (*mat*) and a feature set (*pathways*). To create the feature set, the pathway list from the Kyoto Encyclopedia of Genes and Genomes (KEGG) database, the pathway list from the Reactome database and the pathway list from ECM-related mechanisms (NABA) are loaded. Further, to set up the evaluation matrix, the PKN from COSMOS is loaded, PKN nodes that are not present in the pathways removed and metabolic interactions discarded. Then, by associating all nodes from the PKN/KEGG/Reactome/NABA pathways present in the inferred biological network with value 1 and all nodes not present with value 0, we can create the matrix.

```{r Loading packages, message=FALSE}
library(cosmosR)
library(MOFA2)
library(readr)
library(ggplot2)
library(ggfortify)
library(dplyr)
library(reshape2)
library(liana)
library(decoupleR)
library(moon)
library(pheatmap)
library(gridExtra)
library(liana)
library(GSEABase)
library(tidyr)
library(purrr)
```

```{r Load COSMOS network}
combined_SIF_reduced <- read_csv(file = "results/cosmos/SIF_mofamoon_combined_reduced.csv")
combined_ATT_reduced <- read_csv(file = "results/cosmos/ATT_mofamoon_combined_reduced.csv")
```

```{r import_gmt function, include=FALSE}
import_gmt <- function(gmtfile, fast = T){
  if(fast)
  {
    genesets = GSEABase::getGmt(con = gmtfile)
    genesets = unlist(genesets)
    
    gene_to_term =plyr::ldply(genesets,function(geneset){
      temp <- geneIds(geneset)
      temp2 <- setName(geneset)
      temp3 <- as.data.frame(cbind(temp,rep(temp2,length(temp))))
      
    },.progress = plyr::progress_text())
    names(gene_to_term) <- c("gene","term")
    return(gene_to_term[complete.cases(gene_to_term),])
  }
  else
  {
    genesets = getGmt(con = gmtfile)
    genesets = unlist(genesets)
    
    gene_to_term <- data.frame(NA,NA)
    names(gene_to_term) <- c("gene","term")
    for (geneset in genesets)
    {
      temp <- geneIds(geneset)
      temp2 <- setName(geneset)
      temp3 <- as.data.frame(cbind(temp,rep(temp2,length(temp))))
      names(temp3) <- c("gene","term")
      gene_to_term <- rbind(gene_to_term,temp3)
    }
    
    return(gene_to_term[complete.cases(gene_to_term),])
  }
}
```

```{r ORA analysis with COSMOS network}
## Feature set
pathways_df <- data.frame(import_gmt("support/c2.cp.v2022.1.Hs.symbols.gmt"))

pathways_NABA_KEGG_REACTOME <- data.frame(pathways_df[grepl("NABA_",pathways_df$term) |  grepl("KEGG_",pathways_df$term) | grepl("REACTOME_", pathways_df$term),])
pathways_NABA_KEGG_REACTOME$mor <- 1

## Matrix
data("meta_network")
prots <- unique(c(meta_network$source, meta_network$target))
prots <- gsub("Gene[0-9]+__","",prots)
prots <- gsub("_reverse","",prots)
prots <- prots[!grepl("Metab",prots)]
prots <- prots[prots %in% pathways_NABA_KEGG_REACTOME$gene]

sig_prots <- combined_ATT_reduced$Nodes
sig_prots <- gsub("Gene[0-9]+__","",sig_prots)
sig_prots <- gsub("_reverse","",sig_prots)
sig_prots <- sig_prots[!grepl("Metab",sig_prots)]
sig_prots <- unique(sig_prots)

mat <- matrix(0,nrow = length(prots),1)
row.names(mat) <- prots
mat[row.names(mat) %in% sig_prots, 1] <-  1
mat <- mat[order(mat[,1],decreasing = T), ,drop = F]

## ORA analysis
pathway_ORA_NABA_KEGG_REACTOME_all <- as.data.frame(decoupleR::run_ora(mat = mat, 
                                                     network = pathways_NABA_KEGG_REACTOME, 
                                                     .source = term, 
                                                     .target = gene,
                                                     n_up = length(sig_prots), 
                                                     n_background = length(prots), 
                                                     minsize = 0))

pathway_ORA_NABA_KEGG_REACTOME_all <- 
    pathway_ORA_NABA_KEGG_REACTOME_all %>%
    mutate("adj_score"=-log10(p.adjust(p_value)),
           "start"=0)
```

After the ORA has been performed, the result is returned as a data frame specifying enriched pathways with a p value that is corrected with Benjamini-Hochberg procedure and used to calculate a corresponding adjusted score $(-log_{10}(p)$). We can visualize the results with a ggplot showing the highest scoring pathways.

```{r Visualization of ORA result 1}
pathway_ORA_NABA_KEGG_REACTOME_all <- pathway_ORA_NABA_KEGG_REACTOME_all[order(pathway_ORA_NABA_KEGG_REACTOME_all$adj_score, decreasing = T),]

head(pathway_ORA_NABA_KEGG_REACTOME_all, n = 25)

ggplot(pathway_ORA_NABA_KEGG_REACTOME_all[c(1:10),],aes(adj_score, reorder(source, adj_score, increasing = T))) +
    geom_point(size=5) +
    geom_segment(aes_string(xend = "start", yend = "source")) +
    theme_bw()+
    xlab("Score [-log10(adj_p)]")+
    theme(#axis.text.x = element_blank(), 
        axis.title.y = element_blank(), 
        axis.title.x = element_text(family = "Times New Roman", color = "black"),
        #axis.ticks.x = element_blank(),
        axis.text.y = element_text(family = "Times New Roman", color = "black"),
    )

ggplot(pathway_ORA_NABA_KEGG_REACTOME_all[grepl("NABA", pathway_ORA_NABA_KEGG_REACTOME_all$source),]) +
  geom_point(aes(adj_score, reorder(source, adj_score, increasing = T), color = -log10(p.adjust(p_value))))

ggplot(pathway_ORA_NABA_KEGG_REACTOME_all[grepl("REACTOME", pathway_ORA_NABA_KEGG_REACTOME_all$source) & pathway_ORA_NABA_KEGG_REACTOME_all$adj_score > 30,]) +
  geom_point(aes(adj_score, reorder(source, adj_score, increasing = T), color = -log10(p.adjust(p_value))))

ggplot(pathway_ORA_NABA_KEGG_REACTOME_all[grepl("KEGG", pathway_ORA_NABA_KEGG_REACTOME_all$source) & pathway_ORA_NABA_KEGG_REACTOME_all$adj_score > 15,]) +
  geom_point(aes(adj_score, reorder(source, adj_score, increasing = T), color = -log10(p.adjust(p_value))))
```

Here, for instance, we can see that proteins from pathways in cancer (KEGG) are enriched - in accordance with the fact that we are considering cancer cell lines.

We can further identify and highlight the overlap between nodes present in our network and proteins from the (top) significantly enriched KEGG/NABA/REACTOME pathways.

```{r Visualization of ORA result 2}
sig_pathways <- pathway_ORA_NABA_KEGG_REACTOME_all[pathway_ORA_NABA_KEGG_REACTOME_all$adj_score > 0,"source"]

pathways_adjacency <- dcast(pathways_NABA_KEGG_REACTOME, gene~term)
pathways_adjacency <- pathways_adjacency[pathways_adjacency$gene %in% sig_prots,]
row.names(pathways_adjacency) <- pathways_adjacency$gene
pathways_adjacency <- pathways_adjacency[,sig_pathways]
pathways_adjacency[is.na(pathways_adjacency)] <- 0

pheatmap(t(pathways_adjacency), border_color = T, cluster_rows = F, color = c("lightyellow","red"), breaks = c(0,0.5,1), fontsize = 8, show_rownames = F)

sig_pathways <- pathway_ORA_NABA_KEGG_REACTOME_all[pathway_ORA_NABA_KEGG_REACTOME_all$score > 30,"source"]

pathways_adjacency <- pathways_adjacency[,sig_pathways]

pheatmap(t(pathways_adjacency), border_color = T, cluster_rows = F, color = c("lightyellow","red"), breaks = c(0,0.5,1), fontsize = 7)
```

### ORA with sub-network 

In addition, we can perform an analysis that performs ORA independently for all nodes using the starting node and n-step-distant nodes (in this case 2 steps), enabling us to specifically inspect individual nodes (e.g. SRC) or pathways (e.g. KEGG pathways in cancer).

```{r ORA analysis with sub-networks, eval=FALSE, include=TRUE}
## Matrix
data("meta_network")
prots <- unique(c(meta_network$source, meta_network$target))
prots <- gsub("Gene[0-9]+__","",prots)
prots <- gsub("_reverse","",prots)
prots <- prots[prots %in% pathways_NABA_KEGG_REACTOME$gene]

sig_prots <- combined_SIF_reduced[,c(1,4,2)]
names(sig_prots) <- c("source","interaction","target")
background <- unique(c(sig_prots$source,sig_prots$target))
background_ORA <-  gsub("Gene[0-9]+__","", background)
background_ORA <-  gsub("_reverse","", background_ORA)
background_ORA <- unique(background_ORA)

pathway_NABA_KEGG_REACTOME_list <- list()

for(node in background){

  nodes <- cosmosR:::keep_controllable_neighbours(sig_prots, n_steps = 2, input_nodes = node) #keeps the nodes in network that are no more then n_steps away from the starting nodes
  nodes <- unique(c(nodes$source,nodes$target))
  nodes <- gsub("Gene[0-9]+__","",nodes)
  nodes <- gsub("_reverse","",nodes)
  nodes <- nodes[nodes %in% pathways_NABA_KEGG_REACTOME$gene]
  
  if (length(nodes) > 0) {
    mat <- matrix(0,nrow = length(background_ORA),1)
    row.names(mat) <- background_ORA
    mat[row.names(mat) %in% nodes, 1] <-  1
    mat <- mat[order(mat[,1],decreasing = T), ,drop = F]
    
    pathways_NABA_KEGG_REACTOME_filtered <- pathways_NABA_KEGG_REACTOME %>%
      group_by(term) %>%
      filter(n_distinct(pathways_NABA_KEGG_REACTOME[gene %in% nodes,])>0) %>%
      as.data.frame()
    
    pathway_NABA_KEGG_REACTOME <- as.data.frame(decoupleR::run_ora(mat = mat,
                                                                   network = pathways_NABA_KEGG_REACTOME_filtered,
                                                                   .source = term,
                                                                   .target = gene,
                                                                   n_up = length(nodes),
                                                                   n_background = length(prots),
                                                                   minsize = 0))
    
    pathway_NABA_KEGG_REACTOME <- 
      pathway_NABA_KEGG_REACTOME %>%
      mutate("adj_score"=-log10(p.adjust(p_value)))
    
    pathway_NABA_KEGG_REACTOME <- pathway_NABA_KEGG_REACTOME[,c(6,2)]
    names(pathway_NABA_KEGG_REACTOME) <- c(node,"pathway")
    pathway_NABA_KEGG_REACTOME_list[[node]] <- pathway_NABA_KEGG_REACTOME
    
  } else {

    pathway_NABA_KEGG_REACTOME <-  data.frame(rep(0, length(unique(pathways_NABA_KEGG_REACTOME[pathways_NABA_KEGG_REACTOME$gene %in% background==T,2]))), 
                                                     unique(pathways_NABA_KEGG_REACTOME[pathways_NABA_KEGG_REACTOME$gene %in% background==T,2]))
    names(pathway_NABA_KEGG_REACTOME) <- c(node,"pathway")
    pathway_NABA_KEGG_REACTOME_list[[node]] <- pathway_NABA_KEGG_REACTOME
  
    }
}
save(pathway_NABA_KEGG_REACTOME_list, file = "results/pathway_enrichment/pathway_NABA_KEGG_REACTOME_list.RData")
```

```{r Visualization of ORA result 3}
load("results/pathway_enrichment/pathway_NABA_KEGG_REACTOME_list.RData")
pathway_NABA_KEGG_REACTOME_combined <- data.frame(Reduce(function(x,y) merge(x,y,all.x = T, all.y = T), pathway_NABA_KEGG_REACTOME_list))
row.names(pathway_NABA_KEGG_REACTOME_combined) <- pathway_NABA_KEGG_REACTOME_combined$pathway
pathway_NABA_KEGG_REACTOME_combined <- pathway_NABA_KEGG_REACTOME_combined[,-1]
pathway_NABA_KEGG_REACTOME_combined[is.na(pathway_NABA_KEGG_REACTOME_combined)] <- 0

pheatmap(pathway_NABA_KEGG_REACTOME_combined, fontsize = 6, cluster_rows = T, show_rownames = F)
```

Since cancer cell lines are considered here, we can specifically look at known over-activated cancer pathways.

```{r Visualization of ORA result 4}
cancer_control <- pathway_NABA_KEGG_REACTOME_combined["KEGG_MAPK_SIGNALING_PATHWAY",]
cancer_control <- t(cancer_control[,order(cancer_control[1,], decreasing = F)])
ggplot(cancer_control)+geom_point(aes(reorder(rownames(cancer_control), KEGG_MAPK_SIGNALING_PATHWAY, increasing = T), KEGG_MAPK_SIGNALING_PATHWAY))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

cancer_control2 <- as.data.frame(cancer_control[order(cancer_control, decreasing = T),][c(1:10)])
colnames(cancer_control2) <- "KEGG_MAPK_SIGNALING_PATHWAY"
ggplot(cancer_control2)+geom_point(aes(reorder(rownames(cancer_control2), KEGG_MAPK_SIGNALING_PATHWAY, increasing = T), KEGG_MAPK_SIGNALING_PATHWAY))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

Here, we can see that for MAPK signaling pathway the SRC subnetwork had the highest enrichment score. Thus, specific nodes (e.g. SRC) can be inspected to visualize further pathway involvement.

```{r Visualization of ORA result 5}
SRC_pathway <- pathway_NABA_KEGG_REACTOME_list[["SRC"]][order(pathway_NABA_KEGG_REACTOME_list[["SRC"]]$SRC, decreasing = T),][pathway_NABA_KEGG_REACTOME_list[["SRC"]][order(pathway_NABA_KEGG_REACTOME_list[["SRC"]]$SRC, decreasing = T),]$SRC,]

ggplot(SRC_pathway)+geom_point(aes(reorder(pathway, SRC, increasing = T), SRC))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip()

SRC_pathway <- SRC_pathway[!duplicated(SRC_pathway),]
SRC_pathway <- SRC_pathway[order(SRC_pathway$SRC, decreasing = T),]

ggplot(SRC_pathway[c(1:10),])+geom_point(aes(reorder(pathway, SRC, increasing = T), SRC))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip()
```

## MOFA PCGSEA

To compare the ORA result of the final network, we can also perform principal component gene set enrichment analysis ([PCGSEA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4543476/)) on the MOFA factors containing the protein weights as well as the RNA weights. The result can then be inspected as previously. Further information can be found here: [MOFA2 GSEA](https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/GSEA.html).

```{r PC-GSA analysis with MOFA factors + visualization}
pathways_factor <- pathways_NABA_KEGG_REACTOME %>%
    group_by(term) %>%
    pivot_wider(
        names_from = gene,
        values_from = mor) %>%
  as.data.frame()
rownames(pathways_factor) <- pathways_factor$term
pathways_factor <- pathways_factor[,-1]
pathways_factor[is.na(pathways_factor)] <- 0

# Proteomics
model <- load_model('results/mofa/mofa_res_10factor.hdf5')
model_PCGSA <- model

features_names(model_PCGSA)$proteo <- gsub("_proteo","",features_names(model_PCGSA)$proteo)

factor_GSEA_prot <- MOFA2::run_enrichment(model_PCGSA, "proteo", as.matrix(pathways_factor), min.size = 0, factors = "all", sign = "positive", p.adj.method = "fdr") #here as an example positive weights: features with positive weights are “high” in the samples with positive factor values

factor_GSEA_prot$pval.adj[is.na(factor_GSEA_prot$pval.adj)] <- 1
factor_GSEA_prot$set.statistics[is.na(factor_GSEA_prot$set.statistics)] <- 0

plot_enrichment_heatmap(factor_GSEA_prot)
grid.arrange(plot_enrichment(factor_GSEA_prot, factor=1) + ggtitle("Factor1"),
             plot_enrichment(factor_GSEA_prot, factor=2) + ggtitle("Factor2"),
             plot_enrichment(factor_GSEA_prot, factor=3) + ggtitle("Factor3"),
             plot_enrichment(factor_GSEA_prot, factor=4) + ggtitle("Factor4"),
             plot_enrichment(factor_GSEA_prot, factor=5) + ggtitle("Factor5"))
grid.arrange(plot_enrichment(factor_GSEA_prot, factor=6) + ggtitle("Factor6"),
             plot_enrichment(factor_GSEA_prot, factor=7) + ggtitle("Factor7"),
             plot_enrichment(factor_GSEA_prot, factor=8) + ggtitle("Factor8"),
             plot_enrichment(factor_GSEA_prot, factor=9) + ggtitle("Factor9"))

factor_GSEA_prot_df <- data.frame(rownames(factor_GSEA_prot[["set.statistics"]]), factor_GSEA_prot[["set.statistics"]][,4], factor_GSEA_prot$pval.adj[,4])
colnames(factor_GSEA_prot_df) <- c("source", "score", "p_value")

ggplot(factor_GSEA_prot_df[factor_GSEA_prot_df$score > 4,]) +
  geom_point(aes(score, reorder(source, score, increasing = T), color = p_value)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#Transcriptomics
features_names(model_PCGSA)$RNA <- gsub("_RNA","",features_names(model_PCGSA)$RNA)

factor_GSEA_RNA <- MOFA2::run_enrichment(model_PCGSA, "RNA", as.matrix(pathways_factor), min.size = 0, factors = "all", sign = "positive", p.adj.method = "fdr")

factor_GSEA_RNA$pval.adj[is.na(factor_GSEA_RNA$pval.adj)] <- 1
factor_GSEA_RNA$set.statistics[is.na(factor_GSEA_RNA$set.statistics)] <- 0

plot_enrichment_heatmap(factor_GSEA_RNA)
grid.arrange(plot_enrichment(factor_GSEA_RNA, factor=1) + ggtitle("Factor1"),
             plot_enrichment(factor_GSEA_RNA, factor=2) + ggtitle("Factor2"),
             plot_enrichment(factor_GSEA_RNA, factor=3) + ggtitle("Factor3"),
             plot_enrichment(factor_GSEA_RNA, factor=4) + ggtitle("Factor4"),
             plot_enrichment(factor_GSEA_RNA, factor=5) + ggtitle("Factor5"))
grid.arrange(plot_enrichment(factor_GSEA_RNA, factor=6) + ggtitle("Factor6"),
             plot_enrichment(factor_GSEA_RNA, factor=7) + ggtitle("Factor7"),
             plot_enrichment(factor_GSEA_RNA, factor=8) + ggtitle("Factor8"),
             plot_enrichment(factor_GSEA_RNA, factor=9) + ggtitle("Factor9"))

factor_GSEA_RNA_df <- data.frame(rownames(factor_GSEA_RNA[["set.statistics"]]), factor_GSEA_RNA[["set.statistics"]][,4], factor_GSEA_RNA$pval.adj[,4])
colnames(factor_GSEA_RNA_df) <- c("source", "score", "p_value")

ggplot(factor_GSEA_RNA_df[factor_GSEA_RNA_df$score > 7.5,]) +
  geom_point(aes(score, reorder(source, score, increasing = T), color = p_value)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

Interestingly, it becomes evident in the protein view that although Factor 4 explains not the most of the variance in the proteomics data, it explains the variance of the proteins in the ECM-related pathways (NABA pathways) important for cell reorganization in cancer. Moreover, since RNA transcripts are intermediates to proteins, transcriptomics data can potentially lead to identify over-enriched pathways.

## MOFA FGSEA

Another implementation of GSEA is provided by decoupleR run_fgsea. Here, a modified version of this implementation is used to find overenriched pathways based on MOFA transcriptomics weights.

```{r}
run_fgsea_modified <- function (mat, network, .source = .data$source, .target = .data$target,
          times = 100, nproc = 1, seed = 42, minsize = 5, ...) {
  network <- network %>% rename_net({
    {
      .source
    }
  }, {
    {
      .target
    }
  })
  network <- filt_minsize(rownames(mat), network, minsize)
  regulons <- extract_sets(network)
  if (is.null(colnames(mat))) {
    colnames(mat) <- 1:ncol(mat)
  }
  conditions <- colnames(mat) %>% set_names()
  result_df <- map_dfr(.x = conditions, .f = ~{
    stats <- mat[, .x]
    names(stats) <- rownames(mat)
    options <- list(pathways = regulons, stats = stats, nPermSimple = times,
                    nproc = nproc)
    withr::with_seed(seed, {
      result <- suppressWarnings(do.call(what = fgsea::fgsea,
                                         args = options))
    })
  }, .id = "condition")
  result_df %>%
    dplyr::select(pathway, condition, ES, NES, padj, leadingEdge) %>%
    tidyr::pivot_longer(cols = c("ES","NES"), names_to = "statistic", values_to = "score") %>%
    mutate(statistic = if_else(statistic == "ES", "fgsea",  "norm_fgsea"))
}
```

```{r}
fgsea_result <- run_fgsea_modified(
  get_weights(model_PCGSA)$RNA,
  network = pathways_NABA_KEGG_REACTOME,
  .source = term,
  .target = gene,
  times = 1000,
  minsize = 0,
    nproc = 1
)

fgsea_result_activity_norm <- fgsea_result[fgsea_result$statistic=="norm_fgsea",]
fgsea_result_activity_norm$score[is.na(fgsea_result_activity_norm$score)] <- 0
fgsea_result_activity_norm$padj[is.na(fgsea_result_activity_norm$padj)] <- 1
fgsea_result_activity_norm <- fgsea_result_activity_norm[fgsea_result_activity_norm$padj < 0.05,]
fgsea_result_activity_norm <- fgsea_result_activity_norm[,c(1:2,6)] %>%
  group_by(pathway) %>%
  pivot_wider(
    names_from = condition,
    values_from = score) %>%
  as.data.frame() 
rownames(fgsea_result_activity_norm) <- fgsea_result_activity_norm$pathway

## Top 5 up and top 5 down pathways for all factors
fgsea_result_activity_norm_top_enriched <- lapply(c(2:10), function(x) { 
  y <- data.frame("Score"=rbind(fgsea_result_activity_norm[order(fgsea_result_activity_norm[,x],decreasing = T),c(1,x)][1:5,],
                                fgsea_result_activity_norm[order(fgsea_result_activity_norm[,x],decreasing = F),c(1,x)][1:5,]))
  z <- y[,1]
  y <- y[,-1]
  names(y) <- z
  return(y)
})

fgsea_result_activity_norm_top_enriched_df <- data.frame(t(data.table::rbindlist(
  lapply(fgsea_result_activity_norm_top_enriched, function(x) data.table::data.table(t(x))),
  fill = TRUE)))
colnames(fgsea_result_activity_norm_top_enriched_df) <- colnames(fgsea_result_activity_norm[,-1])
fgsea_result_activity_norm_top_enriched_df[is.na(fgsea_result_activity_norm_top_enriched_df)] <- 0

pheatmap(fgsea_result_activity_norm_top_enriched_df[rowSums(fgsea_result_activity_norm_top_enriched_df)!=0,],
         col=colorRampPalette(c("blue","lightgrey", "red"))(n = 100),
         breaks = seq(-3,3,0.06),
         cluster_cols = F,
         cluster_rows = T,
         show_rownames = T,
         fontfamily="Times New Roman",
         fontsize = 6)

## Top 5 up and top 5 down pathways for factor 4

fgsea_result_activity_norm_top_enriched_df$pathway <- rownames(fgsea_result_activity_norm_top_enriched_df)
fgsea_result_activity_norm_top_enriched_df$start <- 0

tmp2 <- data.frame("pathway"=names(fgsea_result_activity_norm_top_enriched[[4]]),
                   "Factor4"=fgsea_result_activity_norm_top_enriched[[4]],
                   "start"=0)
tmp2$pathway <- factor(tmp2$pathway <-  rownames(tmp2), 
                       levels = tmp2$pathway[order(tmp2$Factor4, decreasing = F)])

ggplot(tmp2, aes_string(x = "pathway", y = "Factor4")) + 
  geom_point(size = 5) + 
  scale_color_manual(values = c("black",  "red")) + 
  geom_segment(aes_string(xend = "pathway", yend = "start")) + 
  coord_flip() + 
  theme(axis.text.y = element_text(size = rel(1), hjust = 1, color = "black"), 
        axis.text.x = element_text(size = rel(1.2),vjust = 0.5, color = "black"), 
        axis.title.y = element_blank(), 
        legend.position = "none", 
        panel.background = element_blank())+
  theme_bw() +
  ylab(label="Normalized enrichment score") +
  xlab("Factor 4")+
  theme(axis.text.x = element_text(family = "Times New Roman", color = "black"),
        axis.title.x = element_text(family = "Times New Roman", color = "black"),
        axis.title.y = element_blank(),
        axis.text.y = element_text(family = "Times New Roman", color = "black"))
```
## Further analyses

Further downstream analyses of the network model (e.g. using [CytoScape](https://genome.cshlp.org/content/13/11/2498)) and the MOFA model (e.g. further comparisons before and after COSMOS) can be performed. In order to perform more targeted analyses related to GSA/GSEA and thus reduce the complexity, it is recommended to identify and extract pathways of interest prior to ORA/PCGSEA.

### Session info

```{r Session info}
sessionInfo()
```
